### PR TITLE
ROSAENG-61040 | refactor: remove github.com/dustin/go-humanize

### PR DESCRIPTION
**Pull request by markirish**
## PR Summary

Replace humanize.Bytes with a ~15-line readableBytes helper using SI units (base 1000) matching the original behavior.

## Detailed Description of the Issue

Our ROSA CLI project currently imports several third-party libraries that are only utilized in a single file (e.g., exclusively within a test file or for a single validation function). Relying on entire external dependencies for isolated use cases unnecessarily increases our security attack surface, complicates dependency management, and bloats the project.

This PR is part of that initiative to remove `github.com/dustin/go-humanize`

## Related Issues and PRs

- Jira: [ROSAENG-61040](https://redhat.atlassian.net/browse/ROSAENG-61040)
- Related design/docs: [https://redhat.atlassian.net/wiki/spaces/ROSACLITF/pages/438405639/Third-Party+Dependency+Audit](https://redhat.atlassian.net/wiki/spaces/ROSACLITF/pages/438405639/Third-Party+Dependency+Audit#github.com%2Fdustin%2Fgo-humanize-(removed))

## Type of Change

- [ ] feat - adds a new user-facing capability.
- [ ] fix - resolves an incorrect behavior or bug.
- [ ] docs - updates documentation only.
- [ ] style - formatting or naming changes with no logic impact.
- [x] refactor - code restructuring with no behavior change.
- [ ] test - adds or updates tests only.
- [ ] chore - maintenance work (tooling, housekeeping, non-product code).
- [ ] build - changes build system, packaging, or dependencies for build output.
- [ ] ci - changes CI pipelines, jobs, or automation workflows.
- [ ] perf - improves performance without changing intended behavior.

### Test Steps
1. run `make test`

### Expected Results

All tests pass

## Proof of the Fix

Tests pass

## Breaking Changes
- [x] No breaking changes
- [ ] Yes, this PR introduces a breaking change (describe impact and migration plan below)

## Developer Verification Checklist
- [x] Commit subject/title follows `[JIRA-TICKET] | [TYPE]: <MESSAGE>`.
- [x] PR description clearly explains both **what** changed and **why**.
- [x] Relevant Jira/GitHub issues and related PRs are linked.
- [x] `make install-hooks` has been run in this clone.
- [x] Tests were added/updated where appropriate.
- [x] I manually tested the change.
- [x] `make test` passes.
- [x] `make lint` passes.
- [x] `make rosa` passes.
- [ ] Documentation or repo-local agent guidance was added/updated where appropriate.
- [x] Any risk, limitation, or follow-up work is documented.

[ROSAENG-61040]: https://redhat.atlassian.net/browse/ROSAENG-61040?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

## Summary

* **Bug Fixes**
  * Standardized download progress and completion to display consistent human-readable byte sizes.
  * Improved non-2xx download error messages with clearer multi-line details, including the HTTP status code and requested URL.
  * Refined timeout error formatting for clearer troubleshooting.

* **Tests**
  * Added unit test coverage for byte-size formatting across boundary values, rounding behavior, and very large counts.

### CODE REVIEW

**Comment by amandahla** [comment]
Lets see how my prompt works:

@coderabbitai
Review this PR only as a third-party dependency removal/replacement ([ROSAENG-61040](https://redhat.atlassian.net/browse/ROSAENG-61040) style).
      Be brief. High-signal findings only. No style nits, no optional refactors, no PR-description rewriting.
      1) Identify OLD module removed and REPLACEMENT type: stdlib / existing go.mod dep / new module / inlined code.
      2) Find the PR/commit that originally added OLD; note any warnings/rationale still relevant.
      3) Check OLD (and NEW if any) GitHub README for warnings, deprecation, or “prefer X instead”.
      4) If NEW module: flag security/CVE risk, license conflict with Apache-2.0, archived/deprecated status.
      5) If inlined code: is it safe, Go-modern (1.26), in a package that matches call sites, and does it add new deps? (new deps undermine the goal)
      6) Confirm go.mod/go.sum cleanup and behavior parity for changed call sites.
      7) Prefer stdlib/existing deps over new modules or copies.
      Output: short verdict (approve / needs change) + bullets only for real risks. Draft a 2–4 sentence evaluation note the author can paste into the PR description. Do not edit the PR.

[ROSAENG-61040]: https://redhat.atlassian.net/browse/ROSAENG-61040?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

File: pkg/helper/download/download.go [hunk 1]
```diff
@@ -103,7 +110,8 @@ func formatDownloadError(err error, url string) error {
 
 	// Timeout errors
 	if strings.Contains(errStr, "timeout") || strings.Contains(errStr, "i/o timeout") {
-		return fmt.Errorf("connection timeout while downloading from %s\nPlease check your internet connection and try again", url)
+		return fmt.Errorf("connection timeout while downloading from %s\n"+
+			"Please check your internet connection and try again", url)
```
[the remark this answers is not in the document]
Comment (amandahla) [reply pkg/helper/download/download.go:114]: This is a constant review...I dont think it should be addressed, changing the error message now. And its out of the scope of the PR
@coderabbitai learn

File: pkg/helper/download/download.go [hunk 2]
```diff
@@ -154,5 +162,23 @@ func (wc WriteCounter) PrintProgress() {
...
+	for n := b / unit; n >= unit; n /= unit {
+		div *= unit
+		exp++
+	}
+	val := float64(b) / float64(div)
+	suffix := []string{"kB", "MB", "GB", "TB", "PB", "EB"}[exp]
+	if math.Round(val*10)/10 < 10 {
+		return fmt.Sprintf("%.1f %s", val, suffix)
+	}
+	return fmt.Sprintf("%.0f %s", val, suffix)
```
[the remark this answers is not in the document]
Comment (amandahla) [reply pkg/helper/download/download.go]: @markirish could you address this? Seems a valid finding.
@coderabbitai give us an example/code excerpt

**Comment by markirish** [comment]
> Tip
> 
> For best results, initiate chat on the files or code changes.
> 
> 🧩 Analysis chain
> 🏁 Script executed:
> 
> ```shell
> #!/bin/bash
> set -euo pipefail
> 
> echo '== PR metadata and changed files =='
> gh pr view 3414 --repo openshift/rosa --json number,url,title,baseRefName,headRefName,commits,files --jq '{number,url,title,base:.baseRefName,head:.headRefName,commits:[.commits[]|{oid:.oid,messageHeadline:.messageHeadline}],files:[.files[]|{path:.path,additions:.additions,deletions:.deletions}]}'
> 
> echo '== Diff against PR base =='
> git diff --no-ext-diff --unified=80 "$(git merge-base HEAD origin/$(gh pr view 3414 --repo openshift/rosa --json baseRefName --jq .baseRefName))" HEAD -- go.mod go.sum pkg/helper/download/download.go pkg/helper/download/readable_bytes_test.go
> 
> echo '== Current module references and formatter call sites =='
> rg -n -C 4 'github\.com/dustin/go-humanize|humanize\.Bytes|readableBytes\s*\(' go.mod go.sum pkg/helper/download
> 
> echo '== Go module version and formatting implementation context =='
> rg -n -C 3 '^go\s+[0-9]+\.[0-9]+' go.mod
> ast-grep outline pkg/helper/download/download.go --items all
> sed -n '1,280p' pkg/helper/download/download.go
> sed -n '1,240p' pkg/helper/download/readable_bytes_test.go
> 
> echo '== Introduction history for old module and call site =='
> git log --all --reverse --format='%H%x09%s' -S'github.com/dustin/go-humanize' -- go.mod
> git log --all --reverse --format='%H%x09%s' -S'humanize.Bytes' -- pkg/helper/download/download.go
> 
> echo '== Introduction commit details (first matching commit) =='
> intro="$(git log --all --reverse --format='%H' -S'github.com/dustin/go-humanize' -- go.mod | head -n1)"
> if [ -n "$intro" ]; then
>   git show --no-ext-diff --format=fuller --stat "$intro"
>   git show --no-ext-diff --format= --unified=30 "$intro" -- go.mod go.sum pkg/helper/download/download.go
> fi
> ```
> 
> Length of output: 43316
> 
> 🤖 web_search executed:
> 
> `github.com/dustin/go-humanize README Bytes documentation warnings deprecation prefer alternative`
> 
> Length of output: 6571
> 
> 🏁 Script executed:
> 
> ```shell
> #!/bin/bash
> set -euo pipefail
> 
> base="$(git merge-base HEAD origin/master)"
> intro="a326f9842515a35186631e877cd7d0ac6c674fa3"
> 
> echo '== Introduction commit metadata and message =='
> git show --no-ext-diff --no-patch --format=fuller "$intro"
> git show --no-ext-diff --format= --unified=40 "$intro" -- go.mod go.sum pkg/helper/download/download.go
> 
> echo '== GitHub PRs associated with introduction commit =='
> gh api -H 'Accept: application/vnd.github+json' \
>   "/repos/openshift/rosa/commits/$intro/pulls" \
>   --jq '.[] | {number,title,url:.html_url,body}'
> 
> echo '== Old v1.0.1 formatter implementation from PR base vendor tree =='
> git show "$base:vendor/github.com/dustin/go-humanize/bytes.go" | sed -n '1,240p'
> 
> echo '== Current PR vendor/module cleanup verification =='
> printf 'go.mod/go.sum module matches: '
> if rg -n 'github\.com/dustin/go-humanize' go.mod go.sum; then false; else echo 'none'; fi
> printf 'vendored path exists: '
> if test -e vendor/github.com/dustin/go-humanize; then echo 'yes'; else echo 'no'; fi
> printf 'vendor/modules.txt match: '
> if rg -n 'github\.com/dustin/go-humanize' vendor/modules.txt; then false; else echo 'none'; fi
> 
> echo '== Compare formatter source behavior over representative edge cases =='
> python3 - <<'PY'
> import math
> 
> def old(s):
>     sizes = ["B", "kB", "MB", "GB", "TB", "PB", "EB"]
>     if s < 10:
>         return f"{s} B"
>     e = math.floor(math.log(float(s), 1000))
>     suffix = sizes[int(e)]
>     val = float(s) / math.pow(1000, math.floor(e))
>     return (f"{val:.1f} {suffix}" if val < 10 else f"{val:.0f} {suffix}")
> 
> def new(b):
>     unit = 1000
>     if b < unit:
>         return f"{b} B"
>     div, exp = unit, 0
>     n = b // unit
>     while n >= unit:
>         div *= unit
>         exp += 1
>         n //= unit
>     val = float(b) / float(div)
>     suffix = ["kB", "MB", "GB", "TB", "PB", "EB"][exp]
>     return f"{val:.1f} {suffix}" if round(val * 10) / 10 < 10 else f"{val:.0f} {suffix}"
> 
> cases = [
>     0, 1, 9, 10, 999, 1000, 9999, 10000,
>     999999, 1000000, 9999999, 10000000,
>     999999999, 1000000000, (2**64)-1
> ]
> for value in cases:
>     print(f"{value}: old={old(value)!r} new={new(value)!r} {'OK' if old(value) == new(value) else 'DIFF'}")
> PY
> ```
> 
> Length of output: 33847
> 
> `@amandahla` **Verdict: needs change**
> 
> * **Behavior parity:** `github.com/dustin/go-humanize` is fully removed (including `go.mod`, `go.sum`, and vendored files) and replaced by a package-local, standard-library-only helper—no new dependency. However, at Line 180, the helper changes output at rounding boundaries: `humanize.Bytes(9999)` returns `10.0 kB`, while `readableBytes(9999)` returns `10 kB` (likewise `9999999`: `10.0 MB` → `10 MB`). The new tests codify this difference, so this is not a byte-for-byte replacement of the only changed call site’s CLI output.
> * **Provenance/upstream:** The module was added for the download command in commit `a326f9842515a35186631e877cd7d0ac6c674fa3` / PR `#68`; no relevant warning or special rationale was found. Its upstream docs define `Bytes` as SI/base-1000 formatting and show no deprecation or “prefer X” guidance.
> 
> **Paste-ready evaluation note:**
> 
> > This change removes the direct `github.com/dustin/go-humanize` dependency and its vendored source, replacing the single `humanize.Bytes` call with a package-local helper using only the Go standard library. The replacement preserves SI/base-1000 units and adds no dependencies. Before merging, the helper should retain the original formatter’s rounding-boundary output (for example, `9999` bytes remains `10.0 kB`) to maintain CLI behavior parity

@coderabbitai Can you confirm that the code now has parity with the old `go-humanize` behavior?

**Comment by markirish** [comment]
https://github.com/openshift/rosa/pull/3414#issuecomment-5071418811

@coderabbitai Ok, check one more time, parity should exist now
